### PR TITLE
Simplify js check in index constraint errors

### LINKED ISSUE #15616: Exception in getSpecialPropertyAssignmentKind when handling `||` in const value
**TypeScript Version:** 2.4.0-dev.20170505 (nightly)

**Code**

```javascript
// (Using react version "15.4.2")
// This is a .js file, using --checkJS (and extra lines were cut out)
define(function(require) {
    'use strict';

    const { Component} = require('react');

    const DEFAULT_MESSAGE = 'Uh-oh, something went wrong...';

    class Errors extends Component {
        _renderMessage(error) {
            const message = error.message || DEFAULT_MESSAGE;
        }
    }
});
```

```bash
# install tsc 20170505 and react 15.4.2 first
tsc --allowJs --checkJs --strict --noEmit example.js
```

**Expected behavior:**
Does not crash
**Actual behavior:**
Crashes because operatorToken is not an object, and the code is trying to access expr.operatorToken.kind

Additionally, with this patch to diagnose the cause, I saw more info

```js
    function getSpecialPropertyAssignmentKind(expression) {
        if (!isInJavaScriptFile(expression)) {
            return 0;
        }     
        var expr = expression;
// begin patch
        if (!expr.operatorToken) {
            console.log('getSpecialPropertyAssignmentKind op', expr);  // This logged, for the '||' operator?
            return 0;
        }     
        if (!expr.left) {
            console.log('getSpecialPropertyAssignmentKind left', expr);
            return 0;
        } 
// end patch
```

Output: https://gist.github.com/TysonAndre/ba9ccf4ada6401d71f877c07f2700dfb

### PR DESCRIPTION
**Pull request by sandersn**
Any declaration that is a BinaryExpression is a special javascript declaration, and all JS declarations that are checked for index constraint compatibility are, in fact, relevant.

Fixes #15616
Improves #15586

### CODE REVIEW

File: tests/cases/compiler/checkIndexConstraintOfJavascriptClassExpression.ts [hunk 1]
```diff
@@ -1,10 +1,18 @@
 // @Filename: weird.js
 // @allowJs: true
+// @checkJs: true
+// @strict: true
+// @noEmit: true
 // @out: foo.js
 someFunction(function(BaseClass) {
+        'use strict';
 	class Hello extends BaseClass {
+                const DEFAULT_MESSAGE = "nop!";
```
Comment (mhegazy) [inline tests/cases/compiler/checkIndexConstraintOfJavascriptClassExpression.ts]: did you mean to move this outside the class?

Comment (sandersn) [inline tests/cases/compiler/checkIndexConstraintOfJavascriptClassExpression.ts]: That does clean up the test output a little. #15616 reproed even with it in the incorrect place, though.

**Comment by mhegazy** [comment]
Please port this to release-2.3

**Comment by sandersn** [comment]
It's now in 2.3 too.

**Pull request merged**